README.md
# LocationsService

Function that returns data for locations by code. Currently only urls are implemented. This retrieves the data set by the LocationsPoller


## Request/Response Structure

Serves `/locations?location_codes=code1,code2...`

Will respond with JSON as below:

```
{
  code1: {
    code: [matching code 1],
    url: [url]
  },
  code2: {
    code: [matching code 2],
    url: [url]
  }
}
```

## Requirements

- ruby 2.7

## Environment variables

- LOG_LEVEL: e.g. 'info' or 'debug', sets the verbosity of the logs

- AWS_REGION: usually 'us-east-1', the region in which to find the locations s3

- BUCKET: the name of the locations s3

- LOCATIONS_FILE: the name of the file in the s3 bucket that contains the locations

- PROFILE: optional, useful for local testing.

## Installation

This function is developed using the AWS SAM framework, [which has installation instructions here](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html)

To install the dependencies for this function, they must be bundled for this framework and should be done with `rake run_bundler`

## Usage

To run the function locally it may be invoked with rake, where FUNCTION is the name of the function you'd like to invoke from the `sam.local.yml` file:

`rake run_local`

## Testing

Testing is provided via `rspec`. The test suite can be invoked with `rake test`


CHANGELOG.md
# CHANGELOG

## 0.0.1 - 2020-10-06
### Added
- Initial commit
- Added support for fetching urls by location code
- Added supporting documentation
- Added unit test suite and basic travis integration


spec/app_spec.rb
require_relative '../app'
require_relative './spec_helper'

ENV['BUCKET'] = 'bucket'
ENV['LOCATIONS_FILE'] = 'locations_file'
ENV['LOG_LEVEL'] = 'info'

describe :init do
  it 'should initialize global variables' do
      mock_s3 = double(Aws::S3::Client)
      mock_response = OpenStruct.new
      mock_response.body = OpenStruct.new
      mock_response.body.string = "{\"ag*\":\"http://fake.com\",\"al*\":\"http://fakefake.com\"}"
      allow(Aws::S3::Client).to receive(:new).and_return(mock_s3)
      allow(mock_s3).to receive(:get_object).and_return(mock_response)
      init
      expect($logger.class).to eq(NYPLRubyUtil::NyplLogFormatter)
      expect($s3_client).to eq(mock_s3)
      expect($locations).to eq(
        {
          /ag.*/=>{:code=>"ag*", :url=>"http://fake.com"},
          /al.*/=>{:code=>"al*", :url=>"http://fakefake.com"}
        }
      )
      expect($initialized).to eq(true)
  end
end

describe :handle_event do
  before(:each) do
    mock_s3 = double(Aws::S3::Client)
    mock_response = OpenStruct.new
    mock_response.body = OpenStruct.new
    mock_response.body.string = "{\"ag*\":\"http://fake.com\",\"al*\":\"http://fakefake.com\"}"
    allow(Aws::S3::Client).to receive(:new).and_return(mock_s3)
    allow(mock_s3).to receive(:get_object).and_return(mock_response)
  end

  it 'should call init' do
    expect(self).to receive(:init)
    handle_event(event: {}, context: {})
  end

  it 'should respond 501 unless it receives a GET request' do
    expect(
      handle_event(event: { 'httpMethod' => 'PUT' }, context: {})
    ).to eq(create_response(501, 'LocationsService only implements GET endpoints'))
  end

  it 'should load_swagger_docs for swagger path' do
    expect(
      handle_event(
        event: {
          'path' => '/docs/locations',
          'httpMethod' => 'GET'
        },
        context: {}
      )
    ).to eq(load_swagger_docs)
  end

  it 'should invoke fetch_locations_and_respond for locations path' do
    expect(self).to receive(:fetch_locations_and_respond)
    handle_event(
      event: {
        'path' => 'base/locations',
        'httpMethod' => 'GET'
      },
      context: {}
    )
  end

  it 'should respond 404 otherwise' do
    expect(
      handle_event(
        event: {
          'httpMethod' => 'GET',
          'path' => 'fake'
        },
        context: {}
      )
    ).to eq(create_response(404, "fake not found"))
  end
end

describe 'fetch_locations_and_respond' do
  it 'should return the correctly mapped locations' do
    expect(
      fetch_locations_and_respond({ 'location_codes' => 'ag,al' })
    ).to eq(
      {
        :body => {
          "ag" => [
            {
              :code => "ag*",
              :url => "http://fake.com"
              }
            ],
          "al" => [
            {
              :code => "al*",
              :url => "http://fakefake.com"
            }
          ]
        },
        :headers => {
          :"Content-type" => "application/json"
        },
        :isBase64Encoded => false,
        :statusCode => 200
      }
    )
  end
end


lib/s3_manager.rb
require 'aws-sdk-s3'

# Class for managing the state of the poller in S3
class S3Client
    # Create S3 client
    def initialize
        @s3 = Aws::S3::Client.new(region: ENV['AWS_REGION'])
    end

    def store_data(file)
        $logger.info "Storing #{file} in s3 for retrieval by check-in card API"

        # Read file into memory
        sql_data = File.read("/tmp/#{file}")

        # Store file in s3
        begin
            @s3.put_object({
                body: sql_data,
                bucket: ENV['SQLITE_BUCKET'],
                key: file,
            })
        rescue StandardError => e
            $logger.error 'Unable to store sqlite db in s3', { status: e.message }
            raise S3Error, 'Unable to store sqlite db in s3'
        end
    end

    def retrieve_data(file)
        $logger.info "Fetching #{file} from s3 bucket"

        begin
            File.open("/tmp/#{file}", 'wb') do |f|
                @s3.get_object(bucket: ENV['SQLITE_BUCKET'], key: file) do |chunk|
                    f.write(chunk)
                end
            end
        rescue StandardError => e
            $logger.error 'Unable to rertrieve file from s3', { status: e.message }
            raise S3Error, 'Unable to retrieve file from s3'
        end
    end
end

class S3Error < StandardError; end


app.rb
require 'nypl_ruby_util'
require 'aws-sdk-s3'
require 'json'

def init
  return if $initialized

  $logger = NYPLRubyUtil::NyplLogFormatter.new($stdout, level: ENV['LOG_LEVEL'])
  s3_config = { region: ENV['AWS_REGION'] }
  s3_config[:profile] = ENV['PROFILE'] if ENV['PROFILE']
  $s3_client = Aws::S3::Client.new(s3_config)

  begin
      raise StandardError.new("missing bucket or locations file") unless ENV['BUCKET'] && ENV['LOCATIONS_FILE']
      locations_response = $s3_client.get_object(bucket: ENV['BUCKET'], key: ENV['LOCATIONS_FILE'])
      $locations = JSON.parse(
        locations_response.body.string
      )
        .map {|k, v| [ Regexp.new(k.gsub('*', '.*')), { code: k, url: v } ]}
        .to_h

  rescue StandardError => e
      $logger.info 'Received s3 error, unable to load locations file from s3', { message: e.message }
      return create_response(500, 'unable to load necessary data from AWS S3')
  end

  $logger.debug 'Initialized function'
  $initialized = true
end

# rubocop:disable Lint/UnusedMethodArgument
def handle_event(event:, context:)
    init

    path = event['path']
    method = event['httpMethod']
    params = event['queryStringParameters']

    $logger.info('handling event', event)

    return create_response(501, 'LocationsService only implements GET endpoints') unless method == 'GET'

    if path == '/docs/locations'
        load_swagger_docs
    elsif /\S+\/locations/.match? path
        fetch_locations_and_respond params
    else
        create_response(404, "#{path} not found")
    end
end

def fetch_locations_and_respond(params)
  req_codes = params['location_codes'].split(",")
  records = req_codes.map do |req_code|
    [
      req_code,
      $locations.select {|k,v| k.match? req_code}.map {|k,v| v}
    ]
  end.to_h
rescue StandardError
    $logger.info 'Received error in fetch_locations_and_respond'
    create_response(500, 'Failed to fetch locations by code')
else
    create_response(200, records)
end

def create_response(status_code = 200, body = nil)
    $logger.info "Responding with #{status_code}"

    {
        statusCode: status_code,
        body: body,
        isBase64Encoded: false,
        headers: { 'Content-type': 'application/json' }
    }
end

def load_swagger_docs
    swagger_docs = JSON.parse(File.read('./swagger.json'))
    create_response(200, swagger_docs)
rescue JSON::JSONError => e
    $logger.error 'Failed to parse Swagger documentation'
    $logger.debug e.message
    create_response(500, 'Unable to load Swagger docs from JSON')
rescue IOError => e
    $logger.error 'Unable to load swagger documentation from file'
    $logger.debug e.message
    create_response(500, 'Unable to load Swagger docs from JSON')
end
